quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,282 @@
+    storage: Arc<dyn Storage>,
+    bundle_file_name: PathBuf,
+    metadata: BundleStorageMetadata,
+}
+
+impl BundleStorage {
+    /// Creates a new BundleStorage.
+    ///
+    /// The provided data must include the footer_bytes at the end of the slice, but it can have more.
+    pub fn new(
+        storage: Arc<dyn Storage>,
+        bundle_file_name: PathBuf,
+        meta_data: &[u8],
+    ) -> io::Result<Self> {
+        let metadata = BundleStorageMetadata::open(meta_data)?;
+        Ok(BundleStorage {
+            storage,
+            bundle_file_name,
+            metadata,
+        })
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct BundleStorageMetadata {
+    pub(crate) files: HashMap<PathBuf, FileOffsets>,
+}
+
+impl BundleStorageMetadata {
+    fn open(data: &[u8]) -> io::Result<Self> {
+        let footer_size: u64 = BinarySerializable::deserialize(&mut &data[data.len() - 8..])?;
+
+        let footer_slice = &data[data.len() - 8 - footer_size as usize..data.len() - 8];
+        serde_json::from_slice(footer_slice)
+            .map_err(|err| io::Error::new(ErrorKind::InvalidData, err))
+    }
+    fn get(&self, path: &Path) -> StorageResult<FileOffsets> {
+        self.files
+            .get(path)
+            .cloned()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "not found in metadata").into())
+    }
+    fn exists(&self, path: &Path) -> bool {
+        self.files.contains_key(path)
+    }
+}
+
+#[async_trait]
+impl Storage for BundleStorage {
+    async fn put(&self, path: &Path, _payload: crate::PutPayload) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn copy_to_file(&self, path: &Path, _output_path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> crate::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
+        Ok(self.metadata.exists(path))
+    }
+
+    fn uri(&self) -> String {
+        self.storage.uri()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+struct FileOffsets {
```

Comment:
Can you add a comment to explain this is a semi-open interval? 
It seems obvious but we've had a bug related to that ambiguity.

Reply:
Alternatively you can use Range<u64>. Apparently it is serializable.

Reply:
I switched to Range, but this is not an semi-open interval, it has start and end

Reply:
Semi-Open or Half-Open means something like this
`start <= ... < end`
In US they write it `[start, end)`. In France it is usually `[start, end[`.

Range in rust is semi-open, while RangeInclusive is closed.

Reply:
I think half-open is when the Range is open on one side, so there is only start or end `(start ..)` or `(..end)`, while `start..end` is a not inclusive closed range